cdap-cli-tests/src/test/java/io/cdap/cdap/cli/CLICommonTests.java

```diff
@@ -0,0 +1,919 @@
+import com.google.common.base.Function;
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.io.Files;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.gson.Gson;
+import io.cdap.cdap.ConfigTestApp;
+import io.cdap.cdap.StandaloneTester;
+import io.cdap.cdap.api.artifact.ArtifactScope;
+import io.cdap.cdap.api.metadata.MetadataEntity;
+import io.cdap.cdap.cli.command.NamespaceCommandUtils;
+import io.cdap.cdap.cli.command.metadata.MetadataCommandHelper;
+import io.cdap.cdap.cli.util.RowMaker;
+import io.cdap.cdap.cli.util.table.Table;
+import io.cdap.cdap.client.DatasetTypeClient;
+import io.cdap.cdap.client.NamespaceClient;
+import io.cdap.cdap.client.ProgramClient;
+import io.cdap.cdap.client.QueryClient;
+import io.cdap.cdap.client.app.FakeApp;
+import io.cdap.cdap.client.app.FakeDataset;
+import io.cdap.cdap.client.app.FakePlugin;
+import io.cdap.cdap.client.app.FakeSpark;
+import io.cdap.cdap.client.app.FakeWorkflow;
+import io.cdap.cdap.client.app.PingService;
+import io.cdap.cdap.client.app.PluginConfig;
+import io.cdap.cdap.client.app.PrefixedEchoHandler;
+import io.cdap.cdap.common.DatasetTypeNotFoundException;
+import io.cdap.cdap.common.io.Locations;
+import io.cdap.cdap.common.test.AppJarHelper;
+import io.cdap.cdap.common.utils.Tasks;
+import io.cdap.cdap.explore.client.ExploreExecutionResult;
+import io.cdap.cdap.proto.DatasetTypeMeta;
+import io.cdap.cdap.proto.NamespaceMeta;
+import io.cdap.cdap.proto.ProgramRunStatus;
+import io.cdap.cdap.proto.ProgramStatus;
+import io.cdap.cdap.proto.QueryStatus;
+import io.cdap.cdap.proto.RunRecord;
+import io.cdap.cdap.proto.WorkflowTokenDetail;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.ArtifactId;
+import io.cdap.cdap.proto.id.DatasetId;
+import io.cdap.cdap.proto.id.DatasetTypeId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.ProgramId;
+import io.cdap.cdap.proto.id.ServiceId;
+import io.cdap.common.cli.CLI;
+import org.apache.twill.filesystem.LocalLocationFactory;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedWriter;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+public class CLICommonTests {
```

Comment:
Instead of defining test methods as static and have other classes calling it, it is better to define this class as a Base class with `@Test` method and have sub-classes to just alter the behavior instead of repeating all the test methods. You can have all the parameters to test methods defined as protected abstract method and let the sub-classes to implement them and return instances needed by the test methods.

Reply:
Done